src/tracing/metric_event.h

```diff
@@ -0,0 +1,181 @@
+        kInstantEvent // No duration
+    };
+
+    // This defines the different types of values that can stored when a metric is emitted
+    struct Value
+    {
+        Value(const Value &)             = default;
+        Value(Value &&)                  = default;
+        Value & operator=(const Value &) = default;
+        Value & operator=(Value &&)      = default;
+
+        enum class Type : uint8_t
+        {
+            kInt32,        // int32_t
+            kUInt32,       // uint32_t
+            kChipErrorCode // chip::ChipError
+        };
+
+        union Store
+        {
+            int32_t int32_value;
+            uint32_t uint32_value;
+
+            Store(int32_t v) : int32_value(v) {}
+
+            Store(uint32_t v) : uint32_value(v) {}
+        };
+
+        Store store;
+        Type type;
+
+        Value(uint32_t value) : store(value), type(Type::kUInt32) {}
+
+        Value(int32_t value) : store(value), type(Type::kInt32) {}
+
+        Value(const ChipError & err) : store(err.AsInteger()), type(Type::kChipErrorCode) {}
+    };
+
+    MetricEvent(Type type, MetricKey key, int32_t value = 0) : mType(type), mKey(key), mValue(value) {}
+
+    MetricEvent(Type type, MetricKey key, uint32_t value) : mType(type), mKey(key), mValue(value) {}
+
+    MetricEvent(Type type, MetricKey key, const ChipError & error) : mType(type), mKey(key), mValue(error) {}
+
+    MetricEvent(Type type, MetricKey key, Value value) : mType(type), mKey(key), mValue(value) {}
+
+    MetricEvent(Type type, MetricKey key, int8_t value) : MetricEvent(type, key, int32_t(value)) {}
+
+    MetricEvent(Type type, MetricKey key, uint8_t value) : MetricEvent(type, key, uint32_t(value)) {}
+
+    MetricEvent(Type type, MetricKey key, int16_t value) : MetricEvent(type, key, int32_t(value)) {}
+
+    MetricEvent(Type type, MetricKey key, uint16_t value) : MetricEvent(type, key, uint32_t(value)) {}
+
+    Type type() const { return mType; }
+
+    MetricKey key() const { return mKey; }
+
+    Value value() const { return mValue; }
+
+    uint32_t ValueUInt32() const { return mValue.store.uint32_value; }
+
+    int32_t ValueInt32() const { return mValue.store.int32_value; }
+
+    uint32_t ValueErrorCode() const { return mValue.store.uint32_value; }
+
+private:
+    Type mType;
+    MetricKey mKey;
+    Value mValue;
+};
+
+namespace utils {
+
+/**
+ * Utility to emit an instant metric if the error is not a success.
+ */
+inline bool LogMetricIfError(MetricKey metricKey, const ::chip::ChipError & err)
+{
+    bool success = ::chip::ChipError::IsSuccess(err);
+    if (!success)
+    {
+        MATTER_LOG_METRIC(metricKey, err);
+    }
+    return success;
+}
+
+/**
+ * This utility class helps generate a Begin and End metric event within the scope of a block using RAII.
+ * This class is also meant to be used in expressions where ChipError object are typically used to capture
+ * error values.
+ */
+class ScopedMetricEvent
```

Comment:
This class does non-trivial work. SHould be unit-testable:
https://github.com/project-chip/connectedhomeip/blob/master/src/tracing/tests/TestTracing.cpp#L55 has examples of unit tests for scoping.